### PR TITLE
Cleanup product interface

### PR DESCRIPTION
**Pull request by mmetayer**
**Description (for Contributor and Core Developer)**

Clean up product interface and its implementation

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php [hunk 1]
```diff
@@ -75,19 +66,19 @@ abstract class AbstractProduct implements ProductInterface
...
-    /** @var ArrayCollection */
+    /** @var Collection */
     protected $uniqueData;
 
-    /** @var ProductModelInterface $parent */
+    /** @var ProductModelInterface|null */
     protected $parent;
 
-    /** @var FamilyVariantInterface */
+    /** @var FamilyVariantInterface|null */
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php:81]: Ideally I would also like to remove this property (the family variant should be defined by the parent IMO), but this has more consequences and would need more tests

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductInterface.php [hunk 2]
```diff
@@ -164,15 +164,6 @@ public function setFamily(FamilyInterface $family = null);
      */
     public function getFamilyId();
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductInterface.php:165]: I chose to keep this method although it's not used (one less BC break), and change its implementation instead

**Review by ahocquard — APPROVED** [review]

**Review by tseho — APPROVED** [review]

**Review by pchasle — APPROVED** [review]
You don't want to add some types on those properties (php7.4 style) while you are in there?

**Comment by mmetayer** [comment]
@pchasle I'm not sure about the impacts yet and I want to keep this PR super light and simple. I might add them in #13178

**Review by nmarniesse — APPROVED** [review]
:+1:

**Pull request merged**